File: packages/core/tests/fixtures/typescript/callbacks.ts
// Callback patterns for testing callback detection

// External callbacks - forEach, map, filter
const callback_numbers = [1, 2, 3, 4, 5];

// Map callback
const doubled = callback_numbers.map((n) => n * 2);

// Filter callback
const evens = callback_numbers.filter((n) => n % 2 === 0);

// ForEach callback
callback_numbers.forEach((n) => console.log(n));

// Internal callbacks - user-defined higher-order function
function run_callback(cb: () => void) {
  cb();
}

run_callback(() => {
  console.log("internal callback");
});

// Nested callbacks
const nested = callback_numbers.map((n) => {
  return [n].filter((x) => x > 2);
});

// Callback with multiple parameters
const pairs = callback_numbers.map((n, i) => ({ value: n, index: i }));

// Non-callback anonymous functions
const standalone_arrow = () => {
  return "not a callback";
};

const standalone_function_expr = function() {
  return "also not a callback";
};

File: packages/core/tests/fixtures/python/callbacks.py
# Callback patterns for testing callback detection

# Lambda callbacks with built-in functions
numbers = [1, 2, 3, 4, 5]

# Map callback
doubled = list(map(lambda x: x * 2, numbers))

# Filter callback
evens = list(filter(lambda x: x % 2 == 0, numbers))

# Sorted with key callback
sorted_by_negative = sorted(numbers, key=lambda x: -x)

# Internal callbacks - user-defined higher-order function
def run_callback(cb):
    cb()

run_callback(lambda: print("internal callback"))

# List comprehension (not a callback - for comparison)
doubled_comp = [x * 2 for x in numbers]
evens_comp = [x for x in numbers if x % 2 == 0]

# Reduce callback (from functools)
from functools import reduce
sum_result = reduce(lambda acc, x: acc + x, numbers, 0)

# Lambda as default argument (not a callback)
def with_default(func=lambda: "default"):
    return func()

# Lambda in variable assignment (not a callback)
standalone_lambda = lambda x: x + 1
